dspy/clients/lm.py

```diff
@@ -151,3 +246,139 @@ def _inspect_history(lm, n: int = 1):
             print(_red(choices_text, end=""))
 
     print("\n\n\n")
+
+
+#-------------------------------------------------------------------------------
+#    Fine-tune support
+#-------------------------------------------------------------------------------
+# TODO: This part can be moved to a separate module
+
+from dspy.clients.openai import (
+    FinetuneJobOpenAI,
+    is_openai_model,
+    finetune_openai,
+)
+from dspy.clients.anyscale import (
+    FinetuneJobAnyScale,
+    finetune_anyscale,
+)
+
+
+_PROVIDER_ANYSCALE = "anyscale"
+_PROVIDER_OPENAI = "openai"
+_SUPPORTED_FINETUNE_PROVIDERS = [
+    _PROVIDER_ANYSCALE,
+    _PROVIDER_OPENAI,
+]
+
+            
+def _get_supported_finetune_provider(model: str) -> Union[str, ValueError]:
+    """Return the finetuning provider for the given model.
+
+    The provider must be in _SUPPORTED_FINETUNE_PROVIDERS. This function is not
+    named `_get_provider` because it does not attempt to find the provider of a
+    model if there is no DSPy fine-tuning support for it.
+    """
+    if is_openai_model(model):
+        return _PROVIDER_OPENAI
+
+    if is_anyscale_model(model):
+        return _PROVIDER_ANYSCALE
+    
+    return ValueError(f"DSPy does not have fine-tuning support for {model}")
+
+
+def get_provider_finetune_job_class(provider: str) -> Type[FinetuneJob]:
+    """Get the FinetuneJob class for the provider."""
+    # Mapping from provider to finetune job type
+    _PROVIDER_TO_FINETUNE_JOB_CLASS = {
+        _PROVIDER_ANYSCALE: FinetuneJobAnyScale,
+        _PROVIDER_OPENAI: FinetuneJobOpenAI,
+    }
+
+    # Get the FinetuneJob class for the provider
+    _CLS = _PROVIDER_TO_FINETUNE_JOB_CLASS[provider]
+
+    return _CLS
+
+
+def get_provider_finetune_function(provider: str) -> callable:
```

Comment:
there's a ton of boilerplate here that definitely shouldn't need to exist IMO

Reply:
Next PR resolves this